### ISSUE TITLE
tilde expansion in path names

### ISSUE THREAD

**Issue by StefanKarpinski** [JuliaLang/julia#1136]
@timholy implemented `tilde_expand` (see [file.jl](https://github.com/JuliaLang/julia/blob/master/base/file.jl#L99)) some time ago and it is now used automatically by a few file-system commands, notably `cd()`. This concerns me, however, since while it's convenient for some things, it's inconsistently used, and it violates the treatment of strings as just data. What if the name of a file is just `~`? We're now in a position where that name needs to be escaped and handled specially in code. On the other hand, tilde expansion is awfully handy.

So here's a possibility: bake tilde expansion into the backtick syntax instead. For example, we could make this work:

``` julia
julia> run(`tail ~/.julia_history`)
           error("~user tilde expansion not yet implemented")
       end
tilde_expand("~")
tilde_expand("~/foo")
tilde_expand("~stefan/foo")
tilde_expand("~stefan")
`cat ~/.julia`
run(`cat ~/.julia`)
run(`tail ~/.julia_history`)
run(`tail $(ENV["HOME"])/.julia_history`)
```

Currently this just looks for a file literally named `~/.julia_history`. Obviously, this would be handy for entering commands, but you could also use it in situations where you want shell-like treatment of any string. For example `cd(`~`)` would expand to `cd(ENV["HOME"])` whereas `cd("~")` would treat its argument literally. Very importantly, tilde expansion would be done at compile time, not based on the run-time data — this is essential to avoiding nasty surprises and corner cases. In particular, interpolated values are not tilde expanded:

``` julia
julia> `~`
`/Users/stefan`

julia> file = "~";

julia> `$file`
`~`
```

One corollary of this approach is that file and directory operations have to work with `Cmd` objects, which maybe ought to be renamed in light of their broadened use-cases under this proposal. Maybe they should be called `ShellString` objects since backtick strings are basically strings with shell-like behavior. Except that they're also not strings, so maybe that's not a great name either.

**Comment by vtjnash** [comment]
since homedir support landed recently in libuv (https://github.com/libuv/libuv/pull/350), we may want to consider switching over to that once https://github.com/JuliaLang/julia/pull/12266 is merged

**Comment by StefanKarpinski** [comment]
Oh nice! We should definitely do that.

**Comment by vtjnash** [comment]
it looks like this proposal would provide a fairly accurate match of the sh semantics (for which tilde expansion precedes variable interpolation). The man page for bash seems to be the most detailed (and the one referenced in the testing above for https://github.com/JuliaLang/julia/issues/1136#issuecomment-7592913). In particular, shells may differ in their handling of `arg=~/path` and `path=$path:~/path`:
```
$ man bash
Tilde Expansion

If a word begins with an unquoted tilde character ('~'), all of the characters preceding the first unquoted slash (or all characters, if there is no unquoted slash) are considered a tilde-prefix. If none of the characters in the tilde-prefix are quoted, the characters in the tilde-prefix following the tilde are treated as a possible login name. If this login name is the null string, the tilde is replaced with the value of the shell parameter HOME. If HOME is unset, the home directory of the user executing the shell is substituted instead. Otherwise, the tilde-prefix is replaced with the home directory associated with the specified login name.

If the tilde-prefix is a '~+', the value of the shell variable PWD replaces the tilde-prefix. If the tilde-prefix is a '~-', the value of the shell variable OLDPWD, if it is set, is substituted. If the characters following the tilde in the tilde-prefix consist of a number N, optionally prefixed by a '+' or a '-', the tilde-prefix is replaced with the corresponding element from the directory stack, as it would be displayed by the dirs builtin invoked with the tilde-prefix as an argument. If the characters following the tilde in the tilde-prefix consist of a number without a leading '+' or '-', '+' is assumed.

If the login name is invalid, or the tilde expansion fails, the word is unchanged.

Each variable assignment is checked for unquoted tilde-prefixes immediately following a : or the first =. In these cases, tilde expansion is also performed. Consequently, one may use file names with tildes in assignments to PATH, MAILPATH, and CDPATH, and the shell assigns the expanded value.
```

My proposal then is to implement this only for a literal `~` inside a backtick at the start of a argument string. Splitting on `=` or `:` would run into platform discrepancies. Thoughts?

**Comment by StefanKarpinski** [comment]
That seems reasonable to me. What are the platform discrepancies you're thinking of for expanding after literal `=` or `:`?

**Comment by vtjnash** [comment]
dos, always dos (aka windows)

**Comment by eschnett** [comment]
So `env "PATH=~/julia/bin:$PATH" julia` would not work from inside Julia?

**Comment by StefanKarpinski** [comment]
What's an example of what might be a bad for Windows here? You mean like `C:~foo` or something? That seems like a pretty unlikely literal. Does the `=` have a special meaning on Windows?

**Comment by vtjnash** [comment]
`=~` is a valid environment variable on windows
`C:~foo` refers to the `~foo` directory in the current working directory on the C drive (and/or relative to the `=C` environment variable)

**Comment by vtjnash** [comment]
looking back over this issue, it's not clear to me that auto-expanding literal `~` is a good idea either. sure, in interactive use it's much more convenient, but i'm concerned that just moves the confusion from the literal to a gotcha when refactoring code into a library.

**Comment by nalimilan** [comment]
In what cases do you expect library code to contain a literal `~`?

**Comment by vtjnash** [comment]
never -- but that's the problem i'm noting (mental disconnect between how you use this at the REPL and how you use it in library code causing unnecessary surprises)

**Comment by nalimilan** [comment]
What I meant is that if you never need to use `~` (or an equivalent way of referring to `$HOME`) in library code, I don't see how expanding it could be an issue.

**Comment by StefanKarpinski** [comment]
In general, I've come to favor making the backtick syntax more shell-like – but with a system-independent set of basic shell features. Features that make sense to me:

- home directory expansion with `~`
- wildcard expansion with `*` and such
- pipeline construction with `|`
- file redirection with `<`, `>` and `>>`

As with other command expansions, they key is that interpretation of the meaning of the command be done before any variable interpolation so that data that happens to appear in a variable cannot affect the meaning of the command, only the data that the command applies to. (This is what other language get tripped up on since they rely on the shell to do the interpretation for them and perform interpolation before passing commands to the shell.)

One thing that gives me pause about this direction is that it might make composition of commands problematic. For example, consider:

```jl
arg = `input.txt | sort >output.txt`
cmd = `head -n1000 $arg`
```

Does this make `cmd` become a pipeline just because a pipeline object was spliced into it? That's a little weird since it looks like you're just splicing a single argument, `arg` (or maybe an array of arguments) into the command. This problem doesn't seem to exist with home directory expansion or glob expansion, both of which still always produce normal command objects. Perhaps the problem could be avoided by disallowing splicing of pipelines or redirected commands into other commands. Then the pipeline construction or file redirection could only occur on the last command, which has a fairly clear and obvious meaning, but you could still build up command arguments incrementally as we allow currently.

**Comment by eschnett** [comment]
Given your starting assumptions, it doesn't make sense to first create a command, and then "extend" it by using it as argument in another command. Strings can be interpolated into a command, but command can only be combined with other commands via pipes, semicolons, parentheses, etc.

You thus write either
```Julia
arg = "input.txt | sort >output.txt"
cmd = `head -n1000 $arg`
```
or
```Julia
arg = `cat input.txt | sort >output.txt`
cmd = pipe(`echo hello`, arg)
```

It is interesting to think about Unix commands that take sub-commands as their argument, e.g. ssh for remote execution. In these cases, it might make sense to say
```Julia
arg = `input.txt | sort >output.txt`
cmd = `ssh $host $arg`
```
This requires great care from Julia to ensure `arg` is quoted correctly, according to the shell that `ssh` will be using to interpret it. However, even here there's no C-macro-style gluing of commands (according to your starting assumptions).

If you really, really want to add a prefix to a command, in the sense of your example, then there's a perfectly good way to do so:
```Julia
arg = prefix -> `$(prefix...) input.txt | sort >output.txt`
cmd = arg("head -n1000")
```
(If the splatting syntax `$(prefix...)` doesn't exist yet then it should be introduced; it's the same syntax as for quoting expressions.)

**Comment by StefanKarpinski** [comment]
@eschnett: the pattern of incrementally constructing commands using interpolation of command objects into other commands is already fairly common and works quite well. It's also common and works well to use command objects to represent the arguments of a command without the command itself and add the command part at the end. This all works because commands are essentially just a way of representing an argument array – and splicing one array into another is a straightforward operation. There's no reason to disallow that.

What is problematic is splicing a pipeline or other more complex construct into a command – the former is no longer as simple as an array of arguments, so splicing it into an array of arguments doesn't make sense. I think that's the only necessary restriction, however. So these usages would be fine:

```jl
cmd = `frizzle`
option1 && (cmd = `$cmd --option1=true`)
option2 && (cmd = `$cmd --option2`)
option3 && (cmd = `$cmd --option3=yes`)
n = parse(Int, readchomp(`$cmd $src $dst | sort -n | head -n1`))

opts = ``
option1 && (opts = `$opts --option1=true`)
option2 && (opts = `$opts --option2`)
option3 && (opts = `$opts --option3=yes`)
args = `$src $dst`
cmd = `frizzle $opts $args`
n = parse(Int, readchomp(`$cmd | sort -n | head -n1`))
```

What would not be ok would be something like this:

```jl
pl = `cat /etc/passwd | head` # OK – returns a pipeline
pl = `$pp -n1` # NOT OK – error
```

**Comment by tkelman** [comment]
this hasn't had any work on it recently, moving to 1.0

**Comment by StefanKarpinski** [comment]
Deprecations in https://github.com/JuliaLang/julia/pull/19786 pave the way for this change in 1.0.

**Comment by StefanKarpinski** [comment]
Seems like we should have this and globbing. Globbing should happen when a command is run, relative to command's working directory. This is actually a feature since you can't use these special characters in command syntax anymore.

**Comment by tecosaur** [comment]
Hmmm, I hope this isn't unwanted, but as a relatively new user, I thought I'd chime in with my confusion that `cd("~/Documents")` didn't work as I expected. I see the comments above that this may be undesirable when referring to files, but whenever a directory is being referend to, wouldn't expanding `~` make sense?

**Comment by goretkin** [comment]
> I see the comments above that this may be undesirable when referring to files, but whenever a directory is being referend to, wouldn't expanding ~ make sense?

There is still the same ambiguity. You could make a directory called `~`:

```julia
julia> cd("/tmp")

julia> mkpath(joinpath(pwd(), "~", "Documents"))
"/private/tmp/~/Documents"

julia> pwd()
"/private/tmp"

julia> cd("~/Documents")

julia> pwd()
"/private/tmp/~/Documents"
```

And that's distinct from
```julia
julia> cd(expanduser("~/Documents"))

julia> pwd()
"/Users/goretkin/Documents"
```

**Comment by ngharrison** [comment]
Just an idea to help users, if the file path doesn't exist and the first character in the string is `~`, add a suggestion to the error message about `~` not being allowed for the home directory.

**Comment by tecosaur** [comment]
That would be great. This does seem like one of a number of spots where the error message could be much more helpful. Rust would probably be a good inspiration in this regard.

**Comment by vtjnash** [comment]
The suggestion has been added to the error message printing for `~` and the cmd parser has its own issue for adding support for more special characters
